File: go/x402gateway/x402.go
package x402gateway

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"io"
	"math"
	"math/big"
	"net/http"
	"regexp"
	"strconv"
	"strings"
	"time"
)

// Method is one way CoinPay can settle under the exact scheme: USDC on one chain.
type Method struct {
	Key     string
	Network string
	Asset   string
	Label   string
}

// Methods lists USDC on Base, Polygon and Ethereum, Base first.
var Methods = []Method{
	{"usdc_base", "eip155:8453", "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913", "USDC on Base"},
	{"usdc_polygon", "eip155:137", "0x3c499c542cEF5E3811e1192ce70d8cC03d5c3359", "USDC on Polygon"},
	{"usdc_eth", "eip155:1", "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48", "USDC on Ethereum"},
}

const decimals = 6

// Domain is the token's EIP-712 domain, the same for all three USDC deployments.
type Domain struct {
	Name    string `json:"name"`
	Version string `json:"version"`
}

// Accept is one entry of an x402 v2 offer.
type Accept struct {
	Scheme            string `json:"scheme"`
	Network           string `json:"network"`
	Amount            string `json:"amount"`
	Asset             string `json:"asset"`
	PayTo             string `json:"payTo"`
	Resource          string `json:"resource"`
	Description       string `json:"description"`
	MimeType          string `json:"mimeType"`
	MaxTimeoutSeconds int    `json:"maxTimeoutSeconds"`
	Extra             Domain `json:"extra"`
}

// Offer is an x402 v2 402 body.
type Offer struct {
	X402Version int      `json:"x402Version"`
	Accepts     []Accept `json:"accepts"`
}

// OfferOptions builds an Offer.
type OfferOptions struct {
	PayTo             string
	PriceCents        float64
	Resource          string
	Description       string
	MaxTimeoutSeconds int // default 300
	Methods           []Method
}

// BuildOffer builds a v2 402 body. Amount is the price in the token's smallest unit, rounded up.
func BuildOffer(o OfferOptions) (Offer, error) {
	if o.PayTo == "" {
		return Offer{}, errors.New("an offer needs a payTo address")
	}
	if o.Description == "" {
		o.Description = "Payment required"
	}
	if o.MaxTimeoutSeconds == 0 {
		o.MaxTimeoutSeconds = 300
	}
	if o.Methods == nil {
		o.Methods = Methods
	}
	amount := strconv.FormatInt(int64(math.Ceil((o.PriceCents/100)*math.Pow10(decimals))), 10)
	out := Offer{X402Version: 2, Accepts: make([]Accept, 0, len(o.Methods))}
	for _, m := range o.Methods {
		out.Accepts = append(out.Accepts, Accept{
			Scheme: "exact", Network: m.Network, Amount: amount, Asset: m.Asset, PayTo: o.PayTo,
			Resource: o.Resource, Description: o.Description, MimeType: "application/json",
			MaxTimeoutSeconds: o.MaxTimeoutSeconds, Extra: Domain{"USD Coin", "2"},
		})
	}
	return out, nil
}

var b64ok = regexp.MustCompile(`^[A-Za-z0-9+/]*={0,2}$`)
var spaces = regexp.MustCompile(`\s+`)

// fromBase64 reads base64 or base64url the forgiving way atob does.
func fromBase64(s string) ([]byte, error) {
	t := spaces.ReplaceAllString(s, "")
	t = strings.NewReplacer("-", "+", "_", "/").Replace(t)
	if len(t)%4 == 0 {
		t = strings.TrimRight(t, "=")
	}
	if len(t)%4 == 1 || !b64ok.MatchString(t) {
		return nil, errors.New("not base64")
	}
	return base64.RawStdEncoding.DecodeString(strings.TrimRight(t, "="))
}

// DecodePayment reads the proof out of an X-PAYMENT header: a JSON object or array, else nil.
func DecodePayment(header string) any {
	if header == "" {
		return nil
	}
	raw, err := fromBase64(header)
	if err != nil {
		return nil
	}
	var v any
	if err := json.Unmarshal(raw, &v); err != nil {
		return nil
	}
	switch v.(type) {
	case map[string]any, []any:
		return v
	}
	return nil
}

// Expected is what CoinPay must hold a proof to, taken from the offered entry for its network.
type Expected struct {
	Amount   string `json:"amount"`
	Resource string `json:"resource"`
	PayTo    string `json:"payTo"`
	Asset    string `json:"asset"`
}

func get(v any, path ...string) any {
	for _, k := range path {
		m, ok := v.(map[string]any)
		if !ok {
			return nil
		}
		v = m[k]
	}
	return v
}

func str(v any) string {
	if v == nil {
		return ""
	}
	switch t := v.(type) {
	case string:
		return t
	case float64:
		return strconv.FormatFloat(t, 'f', -1, 64)
	case bool:
		if t {
			return "true"
		}
		return "false"
	}
	return ""
}

// ExpectedFor finds the offered entry for the proof's network, case-insensitively. Nil if none.
func ExpectedFor(payment any, offer Offer) *Expected {
	network := strings.ToLower(str(get(payment, "network")))
	for _, a := range offer.Accepts {
		if strings.ToLower(a.Network) == network {
			return &Expected{a.Amount, a.Resource, a.PayTo, a.Asset}
		}
	}
	return nil
}

// NonceOf is the single-use nonce CoinPay keys replay detection on.
func NonceOf(payment any) string { return str(get(payment, "payload", "authorization", "nonce")) }

// ValidBeforeOf is when the payer's signature stops being valid, unix seconds; 0 if unreadable.
func ValidBeforeOf(payment any) int64 {
	f, err := strconv.ParseFloat(strings.TrimSpace(str(get(payment, "payload", "authorization", "validBefore"))), 64)
	if err != nil || math.IsInf(f, 0) || math.IsNaN(f) || f <= 0 {
		return 0
	}
	return int64(f)
}

var decimalRe = regexp.MustCompile(`^[+-]?\d+$`)
var hexRe = regexp.MustCompile(`^0[xX][0-9a-fA-F]+$`)

// bigInt reads what BigInt(raw) would: decimal or 0x strings, whole numbers. Nil otherwise.
func bigInt(raw any) *big.Int {
	switch t := raw.(type) {
	case float64:
		if t != math.Trunc(t) {
			return nil
		}
		b, _ := new(big.Float).SetFloat64(t).Int(nil)
		return b
	case string:
		s := strings.TrimSpace(t)
		n := new(big.Int)
		if decimalRe.MatchString(s) {
			n.SetString(strings.TrimPrefix(s, "+"), 10)
			return n
		}
		if hexRe.MatchString(s) {
			n.SetString(s[2:], 16)
			return n
		}
	}
	return nil
}

// PaidValueOf is the value a proof authorizes, in the token's smallest unit, or nil.
func PaidValueOf(payment any) *big.Int {
	raw := get(payment, "payload", "authorization", "value")
	if raw == nil || raw == "" {
		return nil
	}
	v := bigInt(raw)
	if v == nil || v.Sign() <= 0 {
		return nil
	}
	return v
}

// DaysPaid is how many terms value buys at unit per term: a whole number in [1, maxDays], else 0.
func DaysPaid(value *big.Int, unit string, maxDays int) int {
	if value == nil {
		return 0
	}
	per := bigInt(unit)
	if per == nil || per.Sign() <= 0 {
		return 0
	}
	q, r := new(big.Int).QuoRem(value, per, new(big.Int))
	if r.Sign() != 0 || q.Cmp(big.NewInt(1)) < 0 || q.Cmp(big.NewInt(int64(maxDays))) > 0 {
		return 0
	}
	return int(q.Int64())
}

// CoinPay is how the gateway reaches CoinPay's verify and settle routes.
type CoinPay struct {
	APIKey  string
	BaseURL string
	Client  *http.Client
}

func (c CoinPay) post(ctx context.Context, path string, body any) (int, map[string]any) {
	raw, _ := json.Marshal(body)
	ctx, cancel := context.WithTimeout(ctx, 20*time.Second)
	defer cancel()
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.BaseURL+path, bytes.NewReader(raw))
	if err != nil {
		return 0, map[string]any{}
	}
	req.Header.Set("content-type", "application/json")
	req.Header.Set("x-api-key", c.APIKey)
	client := c.Client
	if client == nil {
		client = http.DefaultClient
	}
	res, err := client.Do(req)
	if err != nil {
		return 0, map[string]any{}
	}
	defer res.Body.Close()
	text, _ := io.ReadAll(res.Body)
	var out map[string]any
	if json.Unmarshal(text, &out) != nil || out == nil {
		out = map[string]any{}
	}
	return res.StatusCode, out
}

// Settlement is the outcome of VerifyAndSettle.
type Settlement struct {
	OK     bool
	Payer  string
	Ref    string
	Reason string
	Replay bool
}

var replayVerify = regexp.MustCompile(`(?i)already used|replay`)
var replaySettle = regexp.MustCompile(`(?i)already settled|already being settled`)
var alreadySettled = regexp.MustCompile(`(?i)already settled`)

func truthy(v any) bool {
	switch t := v.(type) {
	case bool:
		return t
	case string:
		return t != ""
	case float64:
		return t != 0
	case nil:
		return false
	}
	return true
}

func firstString(m map[string]any, keys ...string) (string, bool) {
	for _, k := range keys {
		if v, ok := m[k]; ok && v != nil {
			return str(v), true
		}
	}
	return "", false
}

// VerifyAndSettle verifies, then settles. Two calls because verify moves no money.
func (c CoinPay) VerifyAndSettle(ctx context.Context, payment any, expected Expected) Settlement {
	vs, v := c.post(ctx, "/api/x402/verify", map[string]any{"payment": payment, "expected": expected})
	if !truthy(v["valid"]) {
		reason, ok := firstString(v, "error", "reason")
		if !ok {
			reason = "verify failed (" + strconv.Itoa(vs) + ")"
		}
		return Settlement{Reason: reason, Replay: replayVerify.MatchString(reason)}
	}
	ss, s := c.post(ctx, "/api/x402/settle", map[string]any{"payment": payment})
	if !truthy(s["settled"]) {
		reason, ok := firstString(s, "error")
		if !ok {
			reason = "settle failed (" + strconv.Itoa(ss) + ")"
		}
		return Settlement{Reason: reason, Replay: replaySettle.MatchString(reason)}
	}
	ref, ok := firstString(s, "txHash")
	if !ok || ref == "" {
		ref = NonceOf(payment)
	}
	return Settlement{OK: true, Payer: str(get(v, "payment", "from")), Ref: ref}
}

// SettleAgain reports whether the proof has already been paid, when a settle is asked about twice.
func (c CoinPay) SettleAgain(ctx context.Context, payment any) bool {
	_, s := c.post(ctx, "/api/x402/settle", map[string]any{"payment": payment})
	if truthy(s["settled"]) {
		return true
	}
	e, _ := firstString(s, "error")
	return alreadySettled.MatchString(e)
}

File: php/x402-gateway/src/X402.php
<?php
declare(strict_types=1);

namespace Profullstack\X402Gateway;

/** x402 v2 in CoinPay's dialect: the offer, the proof, and the verify/settle calls. */
final class X402
{
    /** USDC on Base, Polygon and Ethereum, Base first. */
    public const METHODS = [
        ['key' => 'usdc_base', 'network' => 'eip155:8453', 'asset' => '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913', 'label' => 'USDC on Base'],
        ['key' => 'usdc_polygon', 'network' => 'eip155:137', 'asset' => '0x3c499c542cEF5E3811e1192ce70d8cC03d5c3359', 'label' => 'USDC on Polygon'],
        ['key' => 'usdc_eth', 'network' => 'eip155:1', 'asset' => '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48', 'label' => 'USDC on Ethereum'],
    ];

    /** A v2 402 body. amount is the price in the token's smallest unit, rounded up. */
    public static function buildOffer(string $payTo, float $priceCents, string $resource, string $description = 'Payment required', int $maxTimeoutSeconds = 300, array $methods = self::METHODS): array
    {
        if ($payTo === '') {
            throw new \InvalidArgumentException('an offer needs a payTo address');
        }
        $amount = (string) (int) ceil(($priceCents / 100) * 1000000);
        $accepts = [];
        foreach ($methods as $m) {
            $accepts[] = [
                'scheme' => 'exact', 'network' => $m['network'], 'amount' => $amount, 'asset' => $m['asset'], 'payTo' => $payTo,
                'resource' => $resource, 'description' => $description, 'mimeType' => 'application/json',
                'maxTimeoutSeconds' => $maxTimeoutSeconds, 'extra' => ['name' => 'USD Coin', 'version' => '2'],
            ];
        }
        return ['x402Version' => 2, 'accepts' => $accepts];
    }

    /** base64 or base64url to text, the forgiving way atob reads it. Null if it is not base64. */
    public static function fromBase64(string $s): ?string
    {
        $t = strtr(preg_replace('/\s+/', '', $s) ?? '', '-_', '+/');
        if (strlen($t) % 4 === 0) {
            $t = rtrim($t, '=');
        }
        if (strlen($t) % 4 === 1 || !preg_match('#^[A-Za-z0-9+/]*={0,2}$#', $t)) {
            return null;
        }
        $t = rtrim($t, '=');
        $out = base64_decode($t . str_repeat('=', (4 - strlen($t) % 4) % 4), true);
        return $out === false ? null : $out;
    }

    /** The proof out of an X-PAYMENT header: an array (object or list), else null. */
    public static function decodePayment(?string $header): ?array
    {
        if ($header === null || $header === '') {
            return null;
        }
        $text = self::fromBase64($header);
        if ($text === null) {
            return null;
        }
        $parsed = json_decode($text, true);
        return is_array($parsed) ? $parsed : null;
    }

    public static function dig(mixed $obj, string ...$path): mixed
    {
        foreach ($path as $k) {
            if (!is_array($obj) || !array_key_exists($k, $obj)) {
                return null;
            }
            $obj = $obj[$k];
        }
        return $obj;
    }

    private static function str(mixed $v): string
    {
        if ($v === null || is_array($v)) {
            return '';
        }
        if (is_bool($v)) {
            return $v ? 'true' : 'false';
        }
        return (string) $v;
    }

    /** What CoinPay must hold the proof to, from the OFFERED entry for its network. */
    public static function expectedFor(mixed $payment, array $offer): ?array
    {
        $network = strtolower(self::str(is_array($payment) ? ($payment['network'] ?? null) : null));
        foreach ($offer['accepts'] ?? [] as $a) {
            if (strtolower($a['network']) === $network) {
                return ['amount' => $a['amount'], 'resource' => $a['resource'], 'payTo' => $a['payTo'], 'asset' => $a['asset']];
            }
        }
        return null;
    }

    public static function nonceOf(mixed $payment): ?string
    {
        $v = self::dig($payment, 'payload', 'authorization', 'nonce');
        return $v === null ? null : self::str($v);
    }

    public static function validBeforeOf(mixed $payment): ?int
    {
        $raw = trim(self::str(self::dig($payment, 'payload', 'authorization', 'validBefore')));
        if ($raw === '' || !is_numeric($raw)) {
            return null;
        }
        $f = (float) $raw;
        return is_finite($f) && $f > 0 ? (int) $f : null;
    }

    /**
     * What BigInt(raw) would read, as a decimal string without sign: decimal or 0x strings, whole
     * numbers. Null when unreadable. Values beyond 64 bits need ext-bcmath or ext-gmp; without
     * either they are refused (null), never mis-read.
     */
    public static function bigint(mixed $raw): ?string
    {
        if (is_bool($raw) || $raw === null || is_array($raw)) {
            return null;
        }
        if (is_int($raw)) {
            return (string) $raw;
        }
        if (is_float($raw)) {
            return is_finite($raw) && floor($raw) === $raw ? number_format($raw, 0, '', '') : null;
        }
        $s = trim((string) $raw);
        if (preg_match('/^[+-]?\d+$/', $s)) {
            $s = ltrim($s, '+');
            if (strlen(ltrim($s, '-')) > 18 && !function_exists('bcadd') && !function_exists('gmp_init')) {
                return null;
            }
            return $s;
        }
        if (preg_match('/^0[xX][0-9a-fA-F]+$/', $s)) {
            $h = substr($s, 2);
            if (strlen($h) <= 15) {
                return (string) hexdec($h);
            }
            if (function_exists('gmp_init')) {
                return gmp_strval(gmp_init($h, 16), 10);
            }
            return null;
        }
        return null;
    }

    private static function cmp(string $a, string $b): int
    {
        if (function_exists('bccomp')) {
            return bccomp($a, $b, 0);
        }
        if (function_exists('gmp_cmp')) {
            return gmp_cmp($a, $b);
        }
        return (int) $a <=> (int) $b;
    }

    /** [quotient, remainder] of two non-negative decimal strings. */
    private static function divmod(string $a, string $b): array
    {
        if (function_exists('bcdiv')) {
            return [bcdiv($a, $b, 0), bcmod($a, $b, 0)];
        }
        if (function_exists('gmp_div_qr')) {
            [$q, $r] = gmp_div_qr($a, $b);
            return [gmp_strval($q), gmp_strval($r)];
        }
        return [(string) intdiv((int) $a, (int) $b), (string) ((int) $a % (int) $b)];
    }

    /** The value a proof authorizes, in the token's smallest unit, or null. */
    public static function paidValueOf(mixed $payment): ?string
    {
        $raw = self::dig($payment, 'payload', 'authorization', 'value');
        if ($raw === null || $raw === '') {
            return null;
        }
        $v = self::bigint($raw);
        return $v !== null && $v[0] !== '-' && self::cmp($v, '0') > 0 ? $v : null;
    }

    /** How many terms $value buys at $unit per term: a whole number in [1, maxDays], else 0. */
    public static function daysPaid(?string $value, string $unit, int $maxDays): int
    {
        if ($value === null) {
            return 0;
        }
        $per = self::bigint($unit);
        if ($per === null || $per[0] === '-' || self::cmp($per, '0') <= 0) {
            return 0;
        }
        [$q, $r] = self::divmod($value, $per);
        if ($r !== '0' || self::cmp($q, '1') < 0 || self::cmp($q, (string) $maxDays) > 0) {
            return 0;
        }
        return (int) $q;
    }

    /** POST JSON. Returns [status, body]. A transport error is [0, '']. */
    public static function post(string $url, array $headers, string $body): array
    {
        $lines = [];
        foreach ($headers as $k => $v) {
            $lines[] = "$k: $v";
        }
        if (function_exists('curl_init')) {
            $ch = curl_init($url);
            curl_setopt_array($ch, [CURLOPT_POST => true, CURLOPT_POSTFIELDS => $body, CURLOPT_HTTPHEADER => $lines, CURLOPT_RETURNTRANSFER => true, CURLOPT_TIMEOUT => 20, CURLOPT_CONNECTTIMEOUT => 20]);
            $text = curl_exec($ch);
            $status = (int) curl_getinfo($ch, CURLINFO_RESPONSE_CODE);
            curl_close($ch);
            return $text === false ? [0, ''] : [$status, (string) $text];
        }
        $ctx = stream_context_create(['http' => ['method' => 'POST', 'header' => implode("\r\n", $lines), 'content' => $body, 'timeout' => 20, 'ignore_errors' => true]]);
        $text = @file_get_contents($url, false, $ctx);
        $status = 0;
        foreach ($http_response_header ?? [] as $h) {
            if (preg_match('#^HTTP/\S+\s+(\d+)#', $h, $m)) {
                $status = (int) $m[1];
            }
        }
        return $text === false ? [0, ''] : [$status, $text];
    }

    private static function obj(string $text): array
    {
        $v = json_decode($text, true);
        return is_array($v) ? $v : [];
    }

    private static function truthy(mixed $v): bool
    {
        return !($v === null || $v === false || $v === '' || $v === 0 || $v === 0.0);
    }

    /** Verify, then settle. ['ok' => true, 'payer', 'ref'] or ['ok' => false, 'reason', 'replay']. */
    public static function verifyAndSettle(mixed $payment, array $expected, string $apiKey, string $baseUrl, callable $post): array
    {
        $call = function (string $path, array $body) use ($apiKey, $baseUrl, $post): array {
            [$status, $text] = $post($baseUrl . $path, ['content-type' => 'application/json', 'x-api-key' => $apiKey], json_encode($body, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE));
            return [$status, self::obj($text)];
        };
        [$vs, $v] = $call('/api/x402/verify', ['payment' => $payment, 'expected' => $expected]);
        if (!self::truthy($v['valid'] ?? null)) {
            $reason = self::str($v['error'] ?? $v['reason'] ?? "verify failed ($vs)");
            return ['ok' => false, 'reason' => $reason, 'replay' => (bool) preg_match('/already used|replay/i', $reason)];
        }
        [$ss, $s] = $call('/api/x402/settle', ['payment' => $payment]);
        if (!self::truthy($s['settled'] ?? null)) {
            $reason = self::str($s['error'] ?? "settle failed ($ss)");
            return ['ok' => false, 'reason' => $reason, 'replay' => (bool) preg_match('/already settled|already being settled/i', $reason)];
        }
        $ref = $s['txHash'] ?? null;
        if ($ref === null || $ref === '') {
            $ref = self::nonceOf($payment);
        }
        return ['ok' => true, 'payer' => self::dig($v, 'payment', 'from'), 'ref' => $ref];
    }

    /** Whether the proof has already been paid, when a settle is asked about twice. */
    public static function settleAgain(mixed $payment, string $apiKey, string $baseUrl, callable $post): bool
    {
        [, $text] = $post("$baseUrl/api/x402/settle", ['content-type' => 'application/json', 'x-api-key' => $apiKey], json_encode(['payment' => $payment], JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE));
        $s = self::obj($text);
        return self::truthy($s['settled'] ?? null) || (bool) preg_match('/already settled/i', self::str($s['error'] ?? ''));
    }
}
